vitrina/datasets/views.py

```diff
@@ -1,16 +1,145 @@
 from django.views.generic import ListView
 from django.views.generic.detail import DetailView
+from django.db.models import Q
 
+from vitrina.datasets.forms import DatasetFilterForm
+from vitrina.helpers import get_selected_value, get_filter_url
 from vitrina.datasets.models import Dataset
+from vitrina.datasets.services import filter_by_status, get_related_categories, get_tag_list, get_related_tag_list, \
+    get_category_counts
+from vitrina.orgs.models import Organization
+from vitrina.classifiers.models import Category
+from vitrina.classifiers.models import Frequency
+from django.http import HttpResponseServerError
 
 
 class DatasetListView(ListView):
     model = Dataset
-    queryset = Dataset.public.order_by('-published')
-    template_name = 'vitrina/datasets/list.html'
     paginate_by = 20
 
+    def get_queryset(self):
+        datasets = Dataset.public.order_by('-published')
+        filter_form = DatasetFilterForm(self.request.GET)
+        if filter_form.is_valid():
+            cleaned_data = filter_form.cleaned_data
+
+            query = cleaned_data.get('q')
+            date_from = cleaned_data.get('date_from')
+            date_to = cleaned_data.get('date_to')
+            status = cleaned_data.get('status')
+            tags = cleaned_data.get('tags')
+            category = cleaned_data.get('category')
+            organization = cleaned_data.get('organization')
+            frequency = cleaned_data.get('frequency')
+
+            if query:
+                datasets = datasets.filter(
+                    Q(title__icontains=query) | Q(title_en__icontains=query)
+                )
+            if date_from and date_to:
+                datasets = datasets.filter(published__range=(date_from, date_to))
+            elif date_from:
+                datasets = datasets.filter(published__gte=date_from)
+            elif date_to:
+                datasets = datasets.filter(published__lte=date_to)
+            if status:
+                datasets = filter_by_status(datasets, status)
+            for tag in tags:
+                datasets = datasets.filter(tags__icontains=tag)
+            if category:
+                related_categories = get_related_categories(category, only_children=True)
+                datasets = datasets.filter(category__pk__in=related_categories)
+            if organization:
+                datasets = datasets.filter(organization=organization)
+            if frequency:
+                datasets = datasets.filter(frequency=frequency)
+            return datasets
+        else:
+            return HttpResponseServerError()
```

Comment:
Please, respect the interface, `get_queryset` must return `QuerySet`, not `HttpResponse`. In order to handle this correctly, you need to move form validation into `get` method where, you can return `HttpResponse`. When you follow the interface, you don't have to resort to hacks like `hasattr(self.object_list, 'status_code')`.

Also, we need to return a [HttpResponseBadRequest](https://docs.djangoproject.com/en/3.2/ref/request-response/#django.http.HttpResponseBadRequest), not a server error.

See also:

- https://github.com/django/django/blob/stable/3.2.x/django/views/generic/list.py#L141-L158
- https://docs.djangoproject.com/en/3.2/ref/class-based-views/generic-display/#listview
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses